core/src/http/cached_resource.rs

```diff
@@ -0,0 +1,188 @@
+//! Fetch HTTP resources, honoring the `Cache-Control` header ([1]) provided by the server.
+//!
+//! [1]: https://datatracker.ietf.org/doc/html/rfc9111#section-5.2
+
+use crate::{
+    http::HttpErrorResponse,
+    retries::retry_http_request,
+    time::{Clock, TimeExt},
+};
+use backoff::{exponential::ExponentialBackoff, SystemClock};
+use http::{header::CACHE_CONTROL, HeaderValue};
+use janus_messages::{Duration, Time};
+use prio::codec::Decode;
+use url::Url;
+
+/// Errors that may arise while managing cached HTTP resources.
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("HTTP client error: {0}")]
+    HttpClient(#[from] reqwest::Error),
+    #[error("codec error: {0}")]
+    Codec(#[from] prio::codec::CodecError),
+    #[error("HTTP response status {0}")]
+    Http(Box<HttpErrorResponse>),
+    #[error("URL parse: {0}")]
+    Url(#[from] url::ParseError),
+    #[error("unexpected server response {0}")]
+    UnexpectedServerResponse(&'static str),
+}
+
+impl From<Result<HttpErrorResponse, reqwest::Error>> for Error {
+    fn from(result: Result<HttpErrorResponse, reqwest::Error>) -> Self {
+        match result {
+            Ok(http_error_response) => Error::Http(Box::new(http_error_response)),
+            Err(error) => error.into(),
+        }
+    }
+}
+
+/// A cached HTTP resource.
+#[derive(Debug, Clone)]
+#[allow(clippy::large_enum_variant)]
+pub enum CachedResource<C, Resource> {
+    Static(Resource),
+    Refreshable(Refresher<C, Resource>),
+}
+
+impl<C: Clock, Resource: Decode> CachedResource<C, Resource> {
+    /// Fetch and cache the resource at the provided URL.
+    pub async fn new(
+        clock: C,
+        resource_url: Url,
+        http_client: &reqwest::Client,
+        http_request_retry_parameters: ExponentialBackoff<SystemClock>,
+    ) -> Result<Self, Error> {
+        let (resource, expires_at) = Refresher::refresh(
+            &clock,
+            http_client,
+            &http_request_retry_parameters,
+            &resource_url,
+        )
+        .await?;
+
+        Ok(Self::Refreshable(Refresher {
+            clock,
+            resource,
+            expires_at,
+            http_client: http_client.clone(),
+            http_request_retry_parameters: http_request_retry_parameters.clone(),
+            resource_url,
+        }))
+    }
+
+    /// Returns the cached resource. Refetches if it has expired.
+    pub async fn resource(&mut self) -> Result<&Resource, Error> {
+        match self {
+            Self::Refreshable(refresher) => refresher.resource().await,
+            Self::Static(ref resource) => Ok(resource),
+        }
+    }
+}
+
+/// Caches an HTTP resource based on the cache-control header provided by the server.
+#[derive(Debug, Clone)]
+pub struct Refresher<C, Resource> {
+    clock: C,
+    resource: Resource,
+    expires_at: Option<Time>,
+    http_client: reqwest::Client,
+    resource_url: Url,
+    http_request_retry_parameters: ExponentialBackoff<SystemClock>,
```

Comment:
These fields could be references, but then the struct would need a generic lifetime parameter, and that would propagate up to `janus_client::Client`, and that would break semver. When I forward-port this to `main`, I'll borrow these fields.